### ISSUE TITLE
Optimize the contract

### ISSUE THREAD

**Issue by gromakovsky** [tqtezos/baseDAO#38]
# Clarification and motivation

In #33 we are already observing that the contract is quite big and in some tests we fail to originate it because we exceed operation size limit (16384 bytes). If we originate it with empty storage, it usually works, but:
1. It works only for trivial DAO where proposal metadata is `()` and instructions in runtime config are trivial.
2. Maybe we'll need to add more functionality later.
3. It's inconvenient that we can't originate the contract with a small (but not empty) storage.
4. Apparently there are some easy ways to make the contract smaller.

# Acceptance criteria

1. Analysis is performed to figure out how the contract can be made smaller.
  * In particular, we'd like to "profile" the contract, i. e. figure out which parts of it are _the most expensive_. Splitting the contract into parts is a bit complicated because for example `creditTo` is used in multiple entrypoints. So it should be just some approximation. We are particularly interested in typechecking gas cost and total size of instructions in bytes. Maybe this profiling will indicate very inefficient places. Even if it doesn't this profiling is still generally useful for us.
2. The contract is made smaller according to the results of that analysis.

**Comment by gromakovsky** [comment]
Some thoughts:

1. `creditFrom` and `debitTo` functions seem to be used 4 and 5 times respectively if my calculations are correct. So one easy optimization that we can make is to move them into storage I suppose. Or not move into storage but put onto stack once at the beginning and take it from stack when necessary. The latter seems better because the originator doesn't need to pass storage with their code.
2. Some errors are quite long, we can save a bit by making them smaller. For example `NOT_ADMINISTRATOR` is used in 8 places and can be changed to `NOT_ADMIN`, saving 64 bytes.
3. Failing with `("ERROR", ())` pair doesn't make much sense because `()` is not informative. We can fail with just `ERROR`. Related `morley` issue that elaborates on that: https://gitlab.com/morley-framework/morley/-/issues/383. I am not sure if it can be done neatly without updating `morley` (and partially doing morley#383), but we can update `morley` if necessary.

**Comment by gromakovsky** [comment]
![image](https://user-images.githubusercontent.com/2556409/99084331-dc59c880-25d7-11eb-9f21-36701c738c28.png)

There is some commented out code in tests. It was commented out because we had to make origination in multiple steps to avoid hitting the operation size limit. After the optimizations within this issue let's restore previous version with single-step origination.

**Comment by Martoon-00** [comment]
> ![image](https://user-images.githubusercontent.com/2556409/99084331-dc59c880-25d7-11eb-9f21-36701c738c28.png)
> 
> There is some commented out code in tests. It was commented out because we had to make origination in multiple steps to avoid hitting the operation size limit. After the optimizations within this issue let's restore previous version with single-step origination.

I suppose I'll create a PR for this after #46 is merged since updating right in #46 seems a bit unrelated.

**Comment by Martoon-00** [comment]
So here goes the profiling report:

https://docs.google.com/spreadsheets/d/16_8XLE-ERIVGqpzbIQDhxeHlPPza7bOpF6JcdTI7FWQ/edit?usp=sharing

From it, one obvious optimization is visible, that we can extract `unfreezeProposerToken` which is used twice in `flush` and save about 1300 bytes in origination operation size. Not sure whether is it worth applying right now though.

**Comment by gromakovsky** [comment]
> I suppose I'll create a PR for this after #46 is merged since updating right in #46 seems a bit unrelated.

Yes, I think this change in tests should belong to #38 and #46 is about a different issue.

**Comment by gromakovsky** [comment]
> From it, one obvious optimization is visible, that we can extract unfreezeProposerToken which is used twice in flush and save about 1300 bytes in origination operation size. Not sure whether is it worth applying right now though.

If it's about 4 hours of work or less (I mean preliminary estimation, we don't know in advance of course), let's do it. Otherwise let's make a new issue and forget about it for now.
